DESIGN.md

```diff
@@ -0,0 +1,279 @@
+which determines what code to use to process incoming messages.
+
+### Message Saver
+
+As a particularly simple example, the fleetspeak demo directory includes a
+[`messagesaver`](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/demo/messagesaver/messagesaver.go)
+package which defines a service which simple saves all received messages to text
+files.
+
+### GRPCService
+
+The
+[`grpcservice`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/server/grpcservice)
+package provides a `service.Factory` which makes a [GRPC](https://grpc.io/) call
+for each delivered message, in this way handing the message off to another
+process.
+
+It expects configuration parameters to be provided in a
+[`fleetspeak.grpcservice.Config`](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/grpcservice/proto/fleetspeak_grpcservice/grpcservice.proto)
+protocol buffer, and the target GRPC server must implement the
+`fleetspeak.grpcservice.Processor` GRPC interface.
+
+In addition to the factory, the grpcservice package also exports a [concrete
+type](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/server/grpcservice#GRPCService)
+which can be used to derive a GRPC-based service with more control over how and
+when it dials a new GRPC target.
+
+
+# Client
+
+The Fleetspeak Client is a small process which runs on the endpoint and
+communicates with a Fleetspeak Server.  Much like the server, it consists of a
+base
+[`client.Client`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client#Client)
+along with a collection of components, and individual installations may adjust
+these components according to specific needs.
+
+## Miniclient: An Example
+
+The
+[`miniclient`](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/demo/miniclient/miniclient.go)
+binary from the demo directory defines a Fleetspeak Client in a general and
+flexible way. Individual installations may want to define their own client
+binary in order to hardcode security critical parameters (for example, before
+signing it as a trusted binary) and otherwise adjust client behavior.
+
+A Fleetspeak client must be provided with a list of TLS root certificates. The
+client will then only communicate a server if it present one of these certs, or
+a cert chaining to one of these certs. The `--trusted_cert_file` flag is
+required by miniclient to set this list.
+
+A Fleetspeek client must also be provided with the server (host and port) to
+connect to. The client will require that the presented TLS server be valid for
+this host. The `--server` flag is required by miniclient to set this server.
+
+In addition, a Fleetspeak client needs a public key that it will use to check
+the signature of externally provided service configurations. The implications of
+this discussed in more detail below. The `--deployment_key_file` flag is
```

Comment:
"are discussed"

Reply:
Done.

Also added the promised discussion to end of Client Services and... section.